include/fuzzuf/algorithms/afl/templates/afl_mutation_hierarflow_routines.hpp

```diff
@@ -1,7 +1,7 @@
 /*
```

Comment:
This file is only formatted, right?

Reply:
There are diffs by adding `havoc_operator_finds` and `last_splice_cycle`.

```diff
diff --git a/old.hpp b/new.hpp
index 54eb619..90b0049 100644
--- a/old.hpp
+++ b/new.hpp
@@ -17,13 +17,12 @@
  */
 #pragma once
 
-#include "fuzzuf/exec_input/exec_input.hpp"
-#include "fuzzuf/utils/common.hpp"
-
 #include "fuzzuf/algorithms/afl/afl_mutator.hpp"
 #include "fuzzuf/algorithms/afl/afl_util.hpp"
-
+#include "fuzzuf/exec_input/exec_input.hpp"
+#include "fuzzuf/optimizer/keys.hpp"
 #include "fuzzuf/optimizer/optimizer.hpp"
+#include "fuzzuf/utils/common.hpp"
 
 namespace fuzzuf::algorithm::afl::routine::mutation {
 
@@ -35,7 +34,6 @@ BitFlip1WithAutoDictBuildTemplate<State>::BitFlip1WithAutoDictBuildTemplate(
 template <class State>
 AFLMutCalleeRef<State> BitFlip1WithAutoDictBuildTemplate<State>::operator()(
     AFLMutatorTemplate<State> &mutator) {
-
   /*********************************************
    * SIMPLE BITFLIP (+dictionary construction) *
    *********************************************/
@@ -166,7 +164,6 @@ AFLMutCalleeRef<State> ByteFlip1WithEffMapBuildTemplate<State>::operator()(
   if (state.eff_cnt != EFF_ALEN<Tag>(mutator.GetLen()) &&
       state.eff_cnt * 100 / EFF_ALEN<Tag>(mutator.GetLen()) >
           option::GetEffMaxPerc(state)) {
-
     std::memset(&state.eff_map[0], 1, EFF_ALEN<Tag>(mutator.GetLen()));
     state.blocks_eff_select += EFF_ALEN<Tag>(mutator.GetLen());
   } else {
@@ -404,6 +401,13 @@ bool HavocBaseTemplate<State>::DoHavoc(
 
   u64 havoc_queued = state.queued_paths;
 
+  std::array<u64, fuzzuf::mutator::NUM_CASE> havoc_operator_finds;
+  bool useHavocOperatorFinds =
+      fuzzuf::optimizer::Store::GetInstance().Exists(
+          fuzzuf::optimizer::keys::SelectedCaseHistogram) &&
+      fuzzuf::optimizer::Store::GetInstance().Exists(
+          fuzzuf::optimizer::keys::SelectedCaseHistogram);
+
   /* We essentially just do several thousand runs (depending on perf_score)
      where we take the input file and make random stacked tweaks. */
 
@@ -418,9 +422,27 @@ bool HavocBaseTemplate<State>::DoHavoc(
     mutator.Havoc(use_stacking, state.extras, state.a_extras, mutop_optimizer,
                   custom_cases);
 
+    u64 havoc_finds = state.queued_paths + state.unique_crashes;
+
     if (this->CallSuccessors(mutator.GetBuf(), mutator.GetLen()))
       return true;
 
+    havoc_finds = state.queued_paths + state.unique_crashes - havoc_finds;
+
+    if (unlikely(havoc_finds > 0) && useHavocOperatorFinds) {
+      auto selected_case_histogram =
+          fuzzuf::optimizer::Store::GetInstance().Get(
+              fuzzuf::optimizer::keys::SelectedCaseHistogram, true);
+      for (size_t i = 0; i < selected_case_histogram.size(); i++) {
+        if (selected_case_histogram[i] > 0) {
+          havoc_operator_finds[i] += havoc_finds;
+        }
+      }
+
+      fuzzuf::optimizer::Store::GetInstance().Set(
+          fuzzuf::optimizer::keys::HavocOperatorFinds, havoc_operator_finds);
+    }
+
     /* out_buf might have been mangled a bit, so let's restore it to its
        original size and shape. */
 
@@ -494,7 +516,6 @@ SplicingTemplate<State>::operator()(AFLMutatorTemplate<State> &mutator) {
   u32 splice_cycle = 0;
   while (splice_cycle++ < option::GetSpliceCycles(state) &&
          state.queued_paths > 1 && mutator.GetSource().GetLen() > 1) {
-
     /* Pick a random queue entry and seek to it. Don't splice with yourself. */
 
     u32 tid;
@@ -544,6 +565,8 @@ SplicingTemplate<State>::operator()(AFLMutatorTemplate<State> &mutator) {
 
     mutator.RestoreSplice();
   }
+  fuzzuf::optimizer::Store::GetInstance().Set(
+      fuzzuf::optimizer::keys::LastSpliceCycle, splice_cycle);
 
   return this->GoToDefaultNext();
 }
```

Reply:
close this as you opened other comments related to this file